features/steps/thens/relations.py

```diff
@@ -131,6 +126,157 @@ def step_impl(context, related, relating, other_entity, condition):
+def step_impl(context, table):
+    if getattr(context, 'applicable', True):
+        tbl_path = system.get_abs_path(f"resources/property_set_definitions/{table}")
+        tbl = system.get_csv(tbl_path, return_type='dict')
+        property_set_definitons = {}
+        for d in tbl:
+            property_set_definitons[d['property_set_name']] = d
+
+        def establish_accepted_pset_values(name, property_set_definitons):
+            def make_obj(s):
+                if s:
+                    return json.loads(s.replace("'", '"'))
+                else:
+                    return ''
+
+            try:
+                property_set_attr = property_set_definitons[name]
+            except KeyError:  # Pset_ not found in template
+                property_set_attr = ''
+                return property_set_attr
+
+            accepted_values = {}
+            accepted_values['template_type'] = property_set_attr.get('template_type', '')
+
+            accepted_values['property_names'] = []
+            accepted_values['property_types'] = []
+            accepted_values['data_types'] = []
+
+            for property_def in make_obj(property_set_attr['property_definitions']):
+                accepted_values['property_names'].append(property_def['property_name'])
+                accepted_values['property_types'].append(property_def['property_type'])
+                accepted_values['data_types'].append(property_def['data_type'])
+
+            accepted_values['applicable_entities'] = make_obj(property_set_attr['applicable_entities'])
+
+            try:
+                accepted_values['applicable_type_values'] = make_obj(property_set_attr.get('applicable_type_value', '')).split(',')
+            except json.decoder.JSONDecodeError:
+                accepted_values['applicable_type_values'] = property_set_attr.get('applicable_type_value', '').split(',')
+
+            return accepted_values
+
+        for inst in context.instances:
+            if isinstance(inst, (tuple, list)):
+                inst = inst[0]
+
+            name = getattr(inst, 'Name', 'Attribute not found')
+
+
+            if 'IfcPropertySet Name attribute value must use predefined values according' in context.step.name:
+                if name not in property_set_definitons.keys():
+                    yield (err.InvalidValueError(False, inst, 'Name', name))  # A custom Pset_ prefixed attribute, e.g. Pset_Mywall
+
+            accepted_values = establish_accepted_pset_values(name, property_set_definitons)
+
+            if accepted_values:  # If not it's a custom Pset_ prefixed attribute, e.g. Pset_Mywall (no need for further Pset_ checks),
+
+                if 'IfcPropertySet must be assigned according to the property set definitions table' in context.step.name:
+                    try:
+                        relations = inst.PropertyDefinitionOf  # IFC2x3
+                    except AttributeError:
+                        relations = inst.DefinesOccurrence # IFC4x3
```

Comment:
I think this should be documented a bit better and linked to the 'Given' statement. 
But, again, this is another discussion related to https://github.com/buildingSMART/ifc-gherkin-rules/pull/89

Also adding these notes a bit for my own reference if you don't mind :)

Do you have more implementations here in the code that is depending of IFC versioning?

Reply:
Here I am not sure what to do. I've added a bit more description, that's a good first step for sure. 

I don't know how to nicely link it to the 'Given' statement, without complicating or duplicating the python function. I still am tempted to keep the current solution, as it's just one try/except block. But I am happy to discuss more, I'm leaving the conversation unresolved for a day or two. 

There are no more places depending on IFC versioning that I'm aware of.

Reply:
I think for now it's more than fine. Thanks for all the helpful comments :-) 

I don't think there's anything you can do. I'm trying to implement a warning when there is another version than IFC4X3_ADD2 (to be edited with newer versions in the future) and uniform the other IFC versions to simplify the versioning, unless there is a rule designed for a specific version. That's why the comments are really helpful.

For the rest the code looks really good. Maybe some minor things, for instance 

 ```
          def make_obj(s):
                if s:
                    return json.loads(s.replace("'", '"'))
                else:
                    return ''
```
Could be simply
          def make_obj(s):
                return json.loads(s.replace("'", '"')) if s else ''